pubsub/rocketmq/rocketmq.go

```diff
@@ -16,120 +16,247 @@ package rocketmq
+				"we will use default model [Clustering]", r.name, r.metadata.ConsumerModel)
+		}
+	}
+	if r.metadata.FromWhere != "" {
+		switch strings.ToLower(r.metadata.FromWhere) {
+		case "consumefromlastoffset":
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromLastOffset))
+		case "consumefromfirstoffset":
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromFirstOffset))
+		case "consumefromtimestamp":
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromTimestamp))
+		default:
+			r.metadata.FromWhere = "ConsumeFromLastOffset"
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromLastOffset))
+			r.logger.Warnf("%s Consumer FromWhere[%s] is error, "+
+				"expected [ConsumeFromLastOffset, ConsumeFromFirstOffset, ConsumeFromTimestamp], "+
+				"we will use default value [ConsumeFromLastOffset]", r.name, r.metadata.FromWhere)
+		}
+	}
+	if r.metadata.ConsumeOrderly != "" {
+		if utils.IsTruthy(r.metadata.ConsumeOrderly) {
+			opts = append(opts, mqc.WithConsumerOrder(true))
+			// in orderly message mode, if no value is set of MessageBatchMaxSize, the recommended value [1] is used
+			if r.metadata.ConsumeMessageBatchMaxSize <= 0 {
+				r.metadata.ConsumeMessageBatchMaxSize = 1
+			}
+		} else {
+			opts = append(opts, mqc.WithConsumerOrder(false))
+		}
+	}
+	if r.metadata.ConsumeMessageBatchMaxSize > 0 {
+		opts = append(opts, mqc.WithConsumeMessageBatchMaxSize(r.metadata.ConsumeMessageBatchMaxSize))
+	}
+	if r.metadata.MaxReconsumeTimes > 0 {
+		opts = append(opts, mqc.WithMaxReconsumeTimes(r.metadata.MaxReconsumeTimes))
+	}
+	if r.metadata.AutoCommit != "" {
+		if utils.IsTruthy(r.metadata.AutoCommit) {
+			opts = append(opts, mqc.WithAutoCommit(true))
+		} else {
+			opts = append(opts, mqc.WithAutoCommit(false))
+		}
+	}
+	if r.metadata.PullInterval > 0 {
+		opts = append(opts, mqc.WithPullInterval(time.Duration(r.metadata.PullInterval)*time.Millisecond))
+	}
+	if r.metadata.PullBatchSize > 0 {
+		opts = append(opts, mqc.WithPullBatchSize(r.metadata.PullBatchSize))
+	}
+	c, e := mqc.NewPushConsumer(opts...)
+	if e != nil {
+		return nil, e
+	}
+	return c, e
 }
 
 func (r *rocketMQ) setUpProducer() (mq.Producer, error) {
 	opts := make([]mqp.Option, 0)
-	if r.metadata.Retries != 0 {
-		opts = append(opts, mqp.WithRetry(r.metadata.Retries))
+	if r.metadata.InstanceName != "" {
+		opts = append(opts, mqp.WithInstanceName(r.metadata.InstanceName))
 	}
-	if r.metadata.GroupName != "" {
-		opts = append(opts, mqp.WithGroupName(r.metadata.GroupName))
+	if r.metadata.ProducerGroupName != "" {
+		opts = append(opts, mqp.WithGroupName(r.metadata.ProducerGroupName))
 	}
-	if r.metadata.ProducerGroup != "" {
-		opts = append(opts, mqp.WithGroupName(r.metadata.ProducerGroup))
-	}
-	if r.metadata.NameServerDomain != "" {
-		opts = append(opts, mqp.WithNameServerDomain(r.metadata.NameServerDomain))
+	if r.metadata.NameServer != "" {
```

Comment:
This is also in `setUpConsumer`: can you please create a helper function for this?

Reply:
ok, i will create a helper function for parse NameServer